CI png optimized check

cx384 opened the pull request:
Adds a workflow to the CI to check whether png files are optimized.
It's only a fuzzy check without `-o7 -zm1-9`, since it would be too slow otherwise, I think, but it should be sufficient.
I hope this makes the review of PRs a little bit easier, and prevents future unoptimized pngs from being added.

Also, it optimizes all existing pngs which have not yet been optimized.
(For some reason, it could save 4.12 KB by optimizing the `cdb_update_cropped.png` file, something most have gone wrong here, since it is only a 16x16 texture.)

## To do

Ready for Review.

## How to test

- Install `optipng` if you don't have it installed
- Run `git ls-files *.png | xargs -I % bash -c "optipng -nc -strip all -simulate % 2>&1 >/dev/null | (if ! grep -q 'is already optimized' $1 ; then echo -e '\033[0;31m%\033[0m'; exit 1; fi)" || (echo "Found non-optimized png file, apply 'optipng -o7 -zm1-9 -nc -strip all -clobber <filename>'"; exit 1;)`
- See how it fails on current master.

sfan5 commented on .github/workflows/png_file_checks.yml:
```diff
@@ -0,0 +1,27 @@
+      # Does only run a fuzzy check without -o7 -zm1-9 since it would be too slow otherwise
+      - name: Check whether all png files are optimized
+        run: |
+          git ls-files *.png | xargs -I % bash -c "optipng -nc -strip all -simulate % 2>&1 >/dev/null | (if ! grep -q 'is already optimized' $1 ; then echo -e '\033[0;31m%\033[0m'; exit 1; fi)" || (echo "Found non-optimized png file, apply 'optipng -o7 -zm1-9 -nc -strip all -clobber <filename>'"; exit 1;)
```
please move this to a script file. there is no need to code-golf this into a single line

cx384 replied:
Done

sfan5 commented on util/ci/check_png_optimized.sh:
```diff
@@ -0,0 +1,18 @@
+
+optimized=0
+
+for file in $(git ls-files *.png); do
```
```suggestion
git ls-files "*.png" | sort -u | while read file; do
```

cx384 replied:
This does not work, since pipes create a subshell where `optimized` is not available anymore, but I applied your suggestion and made it `optimized=$(...)` instead.

I hope this is appropriate.

sfan5 commented on util/ci/check_png_optimized.sh:
```diff
@@ -0,0 +1,18 @@
+
+for file in $(git ls-files *.png); do
+	 # Does only run a fuzzy check without -o7 -zm1-9 since it would be too slow otherwise
+	output=$(optipng -nc -strip all -simulate $file 2>&1 >/dev/null)
```
1. $file should be quoted
2. did you check that it works? you redirect all output to dev null, how could this possibly ever capture anything?

cx384 replied:
1. Changed
2. Yes it works. For some reason, optipng uses stderr for its output message, `2>&1 >/dev/null` redirects stderr to stdout and what would normally be in stdout to /dev/null.
e.g. try `(echo "in"; echo "err" >/dev/stderr) 2>&1 >/dev/null`

sfan5 commented on util/ci/check_png_optimized.sh:
```diff
@@ -0,0 +1,18 @@
+for file in $(git ls-files *.png); do
+	 # Does only run a fuzzy check without -o7 -zm1-9 since it would be too slow otherwise
+	output=$(optipng -nc -strip all -simulate $file 2>&1 >/dev/null)
+	if ! grep -q "is already optimized" <<< $output; then
```
missing quoting

cx384 replied:
I don't think it matters in this case because of `<<<`, but I changed it anyway.

cx384 commented:
I cherry picked the last commit to a `ci_png_failed` branch such that you can see that it works.
https://github.com/cx384/minetest/actions/runs/12467270811/job/34796265368

SmallJoker commented:
A small file size tolerance might be appropriate. Depending on the optimizer, it might differ from the `optipng` output by a few bytes, resulting in unnecessary file changes.

cx384 commented:
Ok, I rewrote the script, now it only fails when it could optimize a png file by more than 3%.
I also adjusted the first commit to only change the more than 3% files.

sfan5 commented on util/ci/check_png_optimized.sh:
```diff
@@ -0,0 +1,31 @@
+	echo "Optimizing png files:"
+	while read file; do
+		# Does only run a fuzzy check without -o7 -zm1-9 since it would be too slow otherwise
+		decrease=($(optipng -nc -strip all -out "$temp_file" -clobber "$file" |& \
```
what does `|&` do? I've never seen that

cx384 replied:
> If `|&` is used, command1’s standard error, in addition to its standard output, is connected to command2’s standard input through the pipe; it is shorthand for `2>&1 |`

See https://www.gnu.org/software/bash/manual/bash.html#Pipelines

(This is needed because optipng outputs to standard error.)

sfan5 reviewed (approved):
Works

cx384 commented:
Rebased to make sure it does not fail on master after merge.
